danceradvent/public/articles/2020/21-asynchronous-intro.pod

```diff
@@ -0,0 +1,590 @@
+
+=back
+
+Over time, we got used to web applications being transaction-based, so
+we created queue managers to help deal with it. We provide all work that
+should be done to the queue manager and it will run it in the background
+while the application continues.
+
+Queue managers are very useful, but they don't solve every situation, and
+in some cases, they are just unnecessary overhead.
+
+=head2 Asynchronous, non-blocking, streaming, what?
+
+There are multiple definitions floating around. There are differences
+between them, but for our purposes, they can all be viewed as similar
+enough to be use interchangeably.
+
+To be a slightly more technical (without the 100 lines of text I have
+decided to spare you), asynchronous code is an umbrella term. Our code
+will write asynchronous code that will work using an event loop and
+non-blocking calls. Streaming is the term for continuously feeding
+a stream of information versus a single response.
+
+=head2 What event loop will we use?
+
+For this example, we will use the L<AnyEvent> event loop. However,
+there are other event loops you should also consider, mainly L<IO::Async>.
+
+The example we use here can be equally written in L<IO::Async> just
+the same. If you have any issues with L<AnyEvent>, we suggest
+researching L<IO::Async>.
+
+The L<Plack> web server we will use with L<AnyEvent> is L<Twiggy>, but
+if you're using L<IO::Async>, you can use L<Net::Async::HTTP::Server>.
+
+=head2 So what example are we using here?
+
+For our example, we built a small application that, when called, will
+reach a Covid-19 API to retrieve the top countries with confirmed cases.
+Then, it will fetch each country's confirmed cases for the last period
+and create a trimmed mean / truncated mean.
+
+=over 4
+
+=item * Why?
+
+This small app provides us with an example of making a single request
+(in our case, an API, but it can be a DB call just the same) and then
+making multiple concurrent requests (again, through an API, but could be
+a DB call).
+
+=item * Could this be cached daily?
+
+Theoretically yes. We're not looking at the most optimal code, but just
+enough contrived code to deliver the message but not entirely useless.
+
+=item * What's a trimmed mean?
+
+Mean (average) is not a very reliable metric, since it can be easily
+offset by outliers. You have ten good grades and one really bad one.
+With mean, you wouldn't look like such a good student. However, if
+we trimmed outliers, we would be able to see you as a good student
+with one crappy grade.
+
+I'm not a statistician, nor am I especially good with math, so if you
+disagree and have a better function, go ahead and use that.
+
+=back
+
+=head2 Where's the code?
+
+=head3 The basics
+
+First, our initial code:
+
+    package CovidStats;
+    use Dancer2;
+
+    # This is still required on the version of Perl we're using
+    # but this won't be "experimental" for much longer
+    use experimental qw< postderef signatures >;
```

Comment:
It might be less confusing if your explicitly require the Perl version where this isn't experimental.